test/replication-luatest/bootstrap_strategy_test.lua

```diff
@@ -179,3 +181,249 @@ g_auto.test_sync_waits_for_all_connected = function(cg)
+end)
+
+local g_config_success = t.group('gh-7999-bootstrap-strategy-config-success', {
+                       {leader = 'server3'},
+                       {leader = uuid3},
+                   })
+
+g_config_success.before_each(function(cg)
+    cg.leader = cg.params.leader
+    -- cg.params can't have "/" for some reason, so recreate the path here.
+    if string.match(cg.leader, 'server3') then
+        cg.leader = server.build_listen_uri(cg.leader)
+    end
+end)
+
+g_config_success.after_each(function(cg)
+    cg.replica_set:drop()
+end)
+
+g_config_success.before_test('test_correct_bootstrap_leader', function(cg)
+    cg.replica_set = replica_set:new{}
+    cg.server1 = cg.replica_set:build_and_add_server{
+        alias = 'server1',
+        box_cfg = {
+            bootstrap_strategy = 'config',
+            bootstrap_leader = cg.leader,
+            instance_uuid = uuid1,
+            replication = {
+                server.build_listen_uri('server1'),
+                server.build_listen_uri('server2'),
+                server.build_listen_uri('server3'),
+            },
+            replication_timeout = 0.1,
+        },
+    }
+    cg.server2 = cg.replica_set:build_and_add_server{
+        alias = 'server2',
+        box_cfg = {
+            replicaset_uuid = uuida,
+            instance_uuid = uuid2,
+        }
+    }
+    cg.server3 = cg.replica_set:build_and_add_server{
+        alias = 'server3',
+        box_cfg = {
+            replicaset_uuid = uuidb,
+            instance_uuid = uuid3,
+            listen = {
+                server.build_listen_uri('server3'),
+            },
+        },
+    }
+end)
+
+g_config_success.test_correct_bootstrap_leader = function(cg)
+    cg.replica_set:start{}
+    t.helpers.retrying({}, cg.server1.exec, cg.server1, function(uuid)
+        local t = require('luatest')
+        t.assert_equals(box.info.cluster.uuid, uuid,
+                        'Server bootstrapped from correct leader')
+    end, {uuidb})
+end
+
+g_config_success.before_test('test_wait_only_for_leader', function(cg)
+    cg.replica_set = replica_set:new{}
+    cg.server1 = cg.replica_set:build_and_add_server{
+        alias = 'server1',
+        box_cfg = {
+            bootstrap_strategy = 'config',
+            bootstrap_leader = cg.leader,
+            replication = {
+                server.build_listen_uri('server1'),
+                server.build_listen_uri('unreachable_2'),
+                server.build_listen_uri('server3'),
+                server.build_listen_uri('unreachable_4'),
+            },
+            replication_connect_timeout = 1000,
+            replication_timeout = 0.1,
+        },
+    }
+    cg.server3 = cg.replica_set:build_and_add_server{
+        alias = 'server3',
+        box_cfg = {
+            replicaset_uuid = uuidb,
+            instance_uuid = uuid3,
+            listen = {
+                server.build_listen_uri('server3'),
+            },
+        },
+    }
+end)
+
```

Comment:
Up to you. I would recommend to cover invalid URIs and UUIDs. Currently all the tests either don't pass them or pass valid values.

Reply:
Indeed, done:
```diff
diff --git a/test/replication-luatest/bootstrap_strategy_test.lua b/test/replication-luatest/bootstrap_strategy_test.lua
index cb894a5fd..ca33c766c 100644
--- a/test/replication-luatest/bootstrap_strategy_test.lua
+++ b/test/replication-luatest/bootstrap_strategy_test.lua
@@ -436,3 +436,22 @@ g_config_success.test_wait_only_for_leader = function(cg)
                         'Server boots as soon as sees the leader')
     end, {uuidb})
 end
+
+local g_config_fail = t.group('gh-7999-bootstrap-strategy-config-fail')
+
+g_config_fail.test_bad_uri_or_uuid = function()
+    local bad_config = {
+        {}, -- empty table.
+        {'a'}, -- non-empty table.
+        {3301},
+        'abracadabra', -- neither a URI or a UUID.
+        'z2345678-1234-1234-1234-12345678', -- not a UUID.
+    }
+    local errmsg = "Incorrect value for option 'bootstrap_leader':"
+    for _, leader in pairs(bad_config) do
+        t.assert_error_msg_contains(errmsg, box.cfg, {
+            bootstrap_strategy = 'config',
+            bootstrap_leader = leader
+        })
+    end
+end 
```